numba/analysis.py

```diff
@@ -434,3 +434,65 @@ def resolve_input_arg_const(input_arg_idx):
     if DEBUG > 1:
         print("after".center(80, '-'))
         print(func_ir.dump())
+
+
+def rewrite_semantic_constants(func_ir, called_args):
+    """
+    This rewrites values known to be constant by their semantics as ir.Const
+    nodes, this is to give branch pruning the best chance possible of killing
+    branches. An example might be rewriting len(tuple) as the literal length.
+
+    func_ir is the IR
+    called_args are the actual arguments with which the function is called
+    """
+    DEBUG = 0
+
+    if DEBUG > 1:
+        print(("rewrite_semantic_constants: " +
+              func_ir.func_id.func_name).center(80, '-'))
+        print("before".center(80, '*'))
+        func_ir.dump()
+
+    def rewrite_statement(func_ir, stmt, new_val):
+        """
+        Rewrites the stmt as a ir.Const new_val and fixes up the entries in
+        func_ir._definitions
+        """
+        stmt.value = ir.Const(new_val, stmt.loc)
+        defns = func_ir._definitions[stmt.target.name]
+        repl_idx = defns.index(val)
+        defns[repl_idx] = stmt.value
+
+    from .ir_utils import get_definition, guard
+    for blk in func_ir.blocks.values():
+        for stmt in blk.body:
+            if isinstance(stmt, ir.Assign):
+                val = stmt.value
+                if isinstance(val, ir.Expr):
+                    if getattr(val, 'op', None) == 'getattr':
```

Comment:
The function has too many levels.  I'd suggest lifting this branch into a `rewrite_array_ndim()` and the next `if` into a `rewrite_tuple_len()`.

Reply:
Good point, done.